caffe2/serialize/intermediate_model.h

```diff
@@ -0,0 +1,399 @@
+    data_ = data;
+  }
+
+  void setStrides(const std::vector<int64_t>& strides) {
+    strides_ = strides;
+  }
+
+  void setDeviceDetail(const IntermediateDeviceOption& device_detail) {
+    deviceDetail_ = device_detail;
+  }
+
+ private:
+  std::string name_;
+  int64_t dataType_;
+  std::vector<int64_t> dims_;
+  int64_t offset_ = 0;
+  std::vector<int64_t> strides_;
+  // TODO: since we still have 2 different Tensor classes in Caffe2 and PyTorch
+  // right now, let's just store the data pointer, and create Tensors
+  // while converting IntermediateModel to the JitScriptModule/Predictor/etc.
+  std::shared_ptr<SharedData> data_;
+  IntermediateDeviceOption deviceDetail_;
+  bool noContent_ = false;
+};
+
+class CAFFE2_API IntermediateParameter final {
+ public:
+  // constructors
+  IntermediateParameter() = default;
+
+  explicit IntermediateParameter(const std::string& name, bool is_buffer, bool require_gradient) :
+    name_(name), isBuffer_(is_buffer), requireGradient_(require_gradient) {}
+
+  explicit IntermediateParameter(torch::ParameterDef* param_def,
+      std::unordered_map<uint64_t, std::shared_ptr<SharedData>>* id_data,
+      DeserializeMode mode);
+
+  // dump data to ParameterDef, invoked in serialize
+  void dump(torch::ParameterDef* param_def);
+
+  // getters/setters
+  const std::string& name() const {
+    return name_;
+  }
+
+  bool isBuffer() const {
+    return isBuffer_;
+  }
+
+  bool requireGradient() const {
+    return requireGradient_;
+  }
+
+  const IntermediateTensor& tensor() const {
+    return tensor_;
+  }
+
+  IntermediateTensor* mutableTensor() {
+    return &tensor_;
+  }
+
+  void setName(const std::string& name) {
+    name_ = name;
+  }
+
+  void setIsBuffer(bool is_buffer) {
+    isBuffer_ = is_buffer;
+  }
+
+  void setRequireGradient(bool require_gradient) {
+    requireGradient_ = require_gradient;
+  }
+
+ private:
+  std::string name_;
+  bool isBuffer_ = false;
+  bool requireGradient_ = false;
+  IntermediateTensor tensor_;
+};
+
+class CAFFE2_API IntermediateMethod final {
+ public:
+  // constructors
+  IntermediateMethod() {};
+
+  explicit IntermediateMethod(torch::MethodDef* method_def);
+
+  IntermediateMethod(const IntermediateMethod& method) = delete;
+  IntermediateMethod& operator =(const IntermediateMethod& method) = delete;
+
+  IntermediateMethod(IntermediateMethod&& method) = default;
+  IntermediateMethod& operator =(IntermediateMethod&& method) = default;
+
+  // dump data to MethodDef, called in serialize
+  void dump(torch::MethodDef* method_def);
+
+  // getters
+  const std::string& name() const {
+    return name_;
+  }
+
+  const std::string& torchScript() const {
+    return torchScript_;
+  }
+
+  // setters
+  void setName(const std::string& name) {
+    name_ = name;
+  }
+
+  void setTorchScript(const std::string& torch_script) {
+    torchScript_ = torch_script;
+  }
+
+ private:
+  std::string name_;
+  //std::unique_ptr<caffe2::NetDef> graph_;
```

Comment:
why did you remove the NetDef? since it's forward declared it should be fine

Reply:
I was doing the experiment :-)